File: app/src/main/java/com/goloveschenko/example/activity/MainActivity.java
package com.goloveschenko.example.activity;

import android.content.Intent;
import android.content.res.Configuration;
import android.support.annotation.IdRes;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;

import com.goloveschenko.example.action.Calculator;
import com.goloveschenko.example.action.Converter;
import com.goloveschenko.example.action.Notation;
import com.goloveschenko.example.action.Operation;
import com.goloveschenko.example.R;
import com.goloveschenko.example.dao.manager.DBManager;
import com.goloveschenko.example.entity.HistoryItem;
import com.goloveschenko.example.fragment.FragmentBin;
import com.goloveschenko.example.fragment.FragmentDec;
import com.goloveschenko.example.fragment.FragmentHex;
import com.goloveschenko.example.fragment.FragmentOct;

import java.math.BigDecimal;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {
    public static final int RESULT_CODE = 200;

    public final static String SYMBOL_PLUS = " + ";
    public final static String SYMBOL_MINUS = " - ";
    public final static String SYMBOL_MULTIPLY = " * ";
    public final static String SYMBOL_DEVIDE = " / ";
    public final static String SYMBOL_POWER = " ^ ";

    public final static String KEY_EXPRESSION = "expression";
    public final static String KEY_NOTATION = "notation";
    public final static String KEY_OPERATION = "operation";
    public final static String KEY_NUMBER = "number";

    private Operation operation;
    private Notation notation;
    private BigDecimal curNumber;
    private EditText inputText;
    private TextView expressionText;

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_history:
                Intent intent = new Intent(MainActivity.this, HistoryActivity.class);
                startActivityForResult(intent, RESULT_CODE);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (data != null) {
            String expr = data.getStringExtra(HistoryActivity.EXTRA_EXPRESSION_RESULT);
            inputText.setText(expr);
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString(KEY_EXPRESSION, expressionText.getText().toString());
        outState.putString(KEY_NOTATION, notation.toString());
        outState.putString(KEY_OPERATION, operation.toString());
        outState.putString(KEY_NUMBER, curNumber.toString());
    }

    @Override
    public void onClick(View v) {
        String text;
        BigDecimal number;

        switch (v.getId()) {
            case R.id.button0:
            case R.id.button1:
            case R.id.button2:
            case R.id.button3:
            case R.id.button4:
            case R.id.button5:
            case R.id.button6:
            case R.id.button7:
            case R.id.button8:
            case R.id.button9:
            case R.id.buttonA:
            case R.id.buttonB:
            case R.id.buttonC:
            case R.id.buttonD:
            case R.id.buttonE:
            case R.id.buttonF:
            case R.id.buttonDevider:
                Button button = (Button) v;
                String expStr;
                if (!inputText.getText().toString().equals("0") || v.getId() == R.id.buttonDevider) {
                    expStr = inputText.getText() + button.getText().toString();
                } else {
                    expStr = button.getText().toString();
                }
                inputText.setText(expStr);
                inputText.setSelection(expStr.length());
                break;
            case R.id.buttonDelete:
                text = inputText.getText().toString();
                if (text.length() > 1) {
                    inputText.setText(text.substring(0, text.length() - 1));
                    inputText.setSelection(text.length() - 1);
                } else {
                    inputText.setText("0");
                    inputText.setSelection(1);
                    curNumber = BigDecimal.ZERO;
                }
                break;
            case R.id.buttonPlus:
                text = inputText.getText().toString();
                if (text.isEmpty()){
                    return;
                }
                expressionText.setText(expressionText.getText() + text + SYMBOL_PLUS);

                number = Converter.stringToValue(notation, text);
                inputText.setText("");
                if (curNumber.compareTo(BigDecimal.ZERO) != 0){
                    curNumber = Calculator.doOperation(operation, curNumber, number);
                } else {
                    curNumber = number;
                }

                operation = Operation.PLUS;
                break;
            case R.id.buttonMinus:
                text = inputText.getText().toString();
                if (text.isEmpty()){
                    return;
                }
                expressionText.setText(expressionText.getText() + text + SYMBOL_MINUS);

                number = Converter.stringToValue(notation, text);
                inputText.setText("");
                if (curNumber.compareTo(BigDecimal.ZERO) != 0){
                    curNumber = Calculator.doOperation(operation, curNumber, number);
                } else {
                    curNumber = number;
                }

                operation = Operation.MINUS;
                break;
            case R.id.buttonMultiply:
                text = inputText.getText().toString();
                if (text.isEmpty()){
                    return;
                }
                expressionText.setText(expressionText.getText() + text + SYMBOL_MULTIPLY);

                number = Converter.stringToValue(notation, text);
                inputText.setText("");
                if (curNumber.compareTo(BigDecimal.ZERO) != 0){
                    curNumber = Calculator.doOperation(operation, curNumber, number);
                } else {
                    curNumber = number;
                }

                operation = Operation.MULTIPLY;
                break;
            case R.id.buttonDevide:
                text = inputText.getText().toString();
                if (text.isEmpty()){
                    return;
                }
                expressionText.setText(expressionText.getText() + text + SYMBOL_DEVIDE);

                number = Converter.stringToValue(notation, text);
                inputText.setText("");
                if (curNumber.compareTo(BigDecimal.ZERO) != 0){
                    curNumber = Calculator.doOperation(operation, curNumber, number);
                } else {
                    curNumber = number;
                }

                operation = Operation.DEVIDE;
                break;
            case R.id.buttonPower:
                text = inputText.getText().toString();
                if (text.isEmpty()){
                    return;
                }
                expressionText.setText(expressionText.getText() + text + SYMBOL_POWER);

                number = new BigDecimal(text);
                inputText.setText("");
                if (curNumber.compareTo(BigDecimal.ZERO) != 0){
                    curNumber = Calculator.doOperation(operation, curNumber, number);
                } else {
                    curNumber = number;
                }

                operation = Operation.POWER;
                break;
            case R.id.buttonResult:
                text = inputText.getText().toString();
                if (text.isEmpty()){
                    return;
                }

                BigDecimal resultNumber;
                if (operation != Operation.NONE) {
                    BigDecimal lastNumber = Converter.stringToValue(notation, text);
                    resultNumber = Calculator.doOperation(operation, curNumber, lastNumber);
                } else {
                    resultNumber = Calculator.parse(text);
                }
                String result = Converter.valueToString(notation, resultNumber);
                inputText.setText(result);

                HistoryItem item = new HistoryItem();
                item.setExpression(expressionText.getText().toString() + text);
                item.setResult(result);
                DBManager.getInstance(getApplicationContext()).insertValue(item);

                curNumber = BigDecimal.ZERO;
                operation = Operation.NONE;
                expressionText.setText("");
                inputText.setSelection(result.length());
                break;
        }
    }

    private void updateInputText(Notation newNotation){
        String oldText = inputText.getText().toString();
        String newText = Converter.convertByNotation(notation, newNotation, oldText);
        inputText.setText(newText);
        inputText.setSelection(inputText.getText().length());
    }

    private void initRadioButtons(){
        if (getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE) {
            RadioGroup radioGroup = (RadioGroup) findViewById(R.id.radioGroup);
            radioGroup.setOnCheckedChangeListener(new RadioGroup.OnCheckedChangeListener() {
                @Override
                public void onCheckedChanged(RadioGroup group, @IdRes int checkedId) {
                    switch (checkedId) {
                        case R.id.radioButtonHex:
                            updateInputText(Notation.HEX);
                            notation = Notation.HEX;
                            break;
                        case R.id.radioButtonDec:
                            updateInputText(Notation.DEC);
                            notation = Notation.DEC;
                            break;
                        case R.id.radioButtonOct:
                            updateInputText(Notation.OCT);
                            notation = Notation.OCT;
                            break;
                        case R.id.radioButtonBin:
                            updateInputText(Notation.BIN);
                            notation = Notation.BIN;
                            break;
                    }
                    initFragment(notation);
                }
            });

            //first checked DEC
            RadioButton radioButton = (RadioButton) findViewById(R.id.radioButtonDec);
            radioButton.setChecked(true);
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        inputText = (EditText) findViewById(R.id.editText);
        expressionText = (TextView) findViewById(R.id.textExpressionView);
        if (savedInstanceState != null) {
            expressionText.setText(savedInstanceState.getString(KEY_EXPRESSION));
            notation = Notation.valueOf(savedInstanceState.getString(KEY_NOTATION));
            operation = Operation.valueOf(savedInstanceState.getString(KEY_OPERATION));
            curNumber = new BigDecimal(savedInstanceState.getString(KEY_NUMBER));
        } else {
            curNumber = BigDecimal.ZERO;
            notation = Notation.DEC;
            operation = Operation.NONE;
        }

        initFragment(Notation.DEC);

        if (getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE) {
            initRadioButtons();
        }

        inputText.setSelection(1);
    }

    private void initFragment(Notation notation) {
        Fragment fragment = null;
        switch (notation) {
            case BIN:
                fragment = new FragmentBin();
                break;
            case OCT:
                fragment = new FragmentOct();
                break;
            case DEC:
                fragment = new FragmentDec();
                break;
            case HEX:
                fragment = new FragmentHex();
                break;
        }
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        transaction.replace(R.id.fragmentCont, fragment)
                .commit();
    }
}

File: app/src/main/java/com/goloveschenko/calculator/activity/MainActivity.kt
package com.goloveschenko.calculator.activity

import android.content.Intent
import android.content.res.Configuration
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Button
import com.goloveschenko.calculator.R
import com.goloveschenko.calculator.action.Calculator
import com.goloveschenko.calculator.action.Converter
import com.goloveschenko.calculator.action.Notation
import com.goloveschenko.calculator.action.Operation
import com.goloveschenko.calculator.dao.manager.DBManager
import com.goloveschenko.calculator.entity.HistoryItem
import com.goloveschenko.calculator.fragment.FragmentBin
import com.goloveschenko.calculator.fragment.FragmentDec
import com.goloveschenko.calculator.fragment.FragmentHex
import com.goloveschenko.calculator.fragment.FragmentOct
import kotlinx.android.synthetic.main.activity_main.*
import java.math.BigDecimal
import java.text.SimpleDateFormat
import java.util.*

class MainActivity : AppCompatActivity(), View.OnClickListener {

    private var operation: Operation = Operation.NONE
    private var notation: Notation = Notation.DEC
    private var curNumber: BigDecimal = BigDecimal.ZERO

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_history -> {
                val intent = Intent(this@MainActivity, HistoryActivity::class.java)
                startActivityForResult(intent, RESULT_CODE)
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (data != null) {
            val expr = data.getStringExtra(HistoryActivity.EXTRA_EXPRESSION_RESULT)
            inputText.setText(expr)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString(KEY_EXPRESSION, expressionText.text.toString())
        outState.putString(KEY_NOTATION, notation.toString())
        outState.putString(KEY_OPERATION, operation.toString())
        outState.putString(KEY_NUMBER, curNumber.toString())
    }

    override fun onClick(v: View) {
        val text: String
        val number: BigDecimal

        when (v.id) {
            R.id.button0,
            R.id.button1,
            R.id.button2,
            R.id.button3,
            R.id.button4,
            R.id.button5,
            R.id.button6,
            R.id.button7,
            R.id.button8,
            R.id.button9,
            R.id.buttonA,
            R.id.buttonB,
            R.id.buttonC,
            R.id.buttonD,
            R.id.buttonE,
            R.id.buttonF,
            R.id.buttonDevider -> {
                val button = v as Button
                val expStr: String
                if (inputText.text.toString() != "0" || v.getId() == R.id.buttonDevider) {
                    expStr = inputText.text.toString() + button.text.toString()
                } else {
                    expStr = button.text.toString()
                }
                inputText.setText(expStr)
                inputText.setSelection(expStr.length)
            }
            R.id.buttonDelete -> {
                text = inputText.text.toString()
                if (text.length > 1) {
                    inputText.setText(text.substring(0, text.length - 1))
                    inputText.setSelection(text.length - 1)
                } else {
                    inputText.setText("0")
                    inputText.setSelection(1)
                    curNumber = BigDecimal.ZERO
                }
            }
            R.id.buttonPlus -> {
                text = inputText.text.toString()
                if (text.isEmpty()) {
                    return
                }
                expressionText.text = expressionText.text.toString() + text + SYMBOL_PLUS

                number = Converter.stringToValue(notation, text)
                inputText.setText("")
                if (curNumber.compareTo(BigDecimal.ZERO) != 0) {
                    curNumber = Calculator.doOperation(operation, curNumber, number)
                } else {
                    curNumber = number
                }

                operation = Operation.PLUS
            }
            R.id.buttonMinus -> {
                text = inputText.text.toString()
                if (text.isEmpty()) {
                    return
                }
                expressionText.text = expressionText.text.toString() + text + SYMBOL_MINUS

                number = Converter.stringToValue(notation, text)
                inputText.setText("")
                if (curNumber.compareTo(BigDecimal.ZERO) != 0) {
                    curNumber = Calculator.doOperation(operation, curNumber, number)
                } else {
                    curNumber = number
                }

                operation = Operation.MINUS
            }
            R.id.buttonMultiply -> {
                text = inputText.text.toString()
                if (text.isEmpty()) {
                    return
                }
                expressionText.text = expressionText.text.toString() + text + SYMBOL_MULTIPLY

                number = Converter.stringToValue(notation, text)
                inputText.setText("")
                if (curNumber.compareTo(BigDecimal.ZERO) != 0) {
                    curNumber = Calculator.doOperation(operation, curNumber, number)
                } else {
                    curNumber = number
                }

                operation = Operation.MULTIPLY
            }
            R.id.buttonDevide -> {
                text = inputText.text.toString()
                if (text.isEmpty()) {
                    return
                }
                expressionText.text = expressionText.text.toString() + text + SYMBOL_DEVIDE

                number = Converter.stringToValue(notation, text)
                inputText.setText("")
                if (curNumber.compareTo(BigDecimal.ZERO) != 0) {
                    curNumber = Calculator.doOperation(operation, curNumber, number)
                } else {
                    curNumber = number
                }

                operation = Operation.DEVIDE
            }
            R.id.buttonPower -> {
                text = inputText.text.toString()
                if (text.isEmpty()) {
                    return
                }
                expressionText.text = expressionText.text.toString() + text + SYMBOL_POWER

                number = BigDecimal(text)
                inputText.setText("")
                if (curNumber.compareTo(BigDecimal.ZERO) != 0) {
                    curNumber = Calculator.doOperation(operation, curNumber, number)
                } else {
                    curNumber = number
                }

                operation = Operation.POWER
            }
            R.id.buttonResult -> {
                text = inputText.text.toString()
                if (text.isEmpty()) {
                    return
                }

                val resultNumber: BigDecimal
                if (operation != Operation.NONE) {
                    val lastNumber = Converter.stringToValue(notation, text)
                    resultNumber = Calculator.doOperation(operation, curNumber, lastNumber)
                } else {
                    resultNumber = Calculator.parse(text)
                }
                val result = Converter.valueToString(notation, resultNumber)
                inputText.setText(result)

                val historyItem = HistoryItem(null,
                        SimpleDateFormat(HistoryActivity.DATE_FORMAT, Locale.US).format(Date()),
                        expressionText.text.toString() + text,
                        result,
                        null)
                DBManager.getInstance(applicationContext).insertValue(historyItem)

                curNumber = BigDecimal.ZERO
                operation = Operation.NONE
                expressionText.text = ""
                inputText.setSelection(result.length)
            }
        }
    }

    private fun updateInputText(newNotation: Notation) {
        val oldText = inputText.text.toString()
        val newText = Converter.convertByNotation(notation, newNotation, oldText)
        inputText.setText(newText)
        inputText.setSelection(inputText.text.length)
    }

    private fun initRadioButtons() {
        if (resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE) {
            radioGroup.setOnCheckedChangeListener { group, checkedId ->
                when (checkedId) {
                    R.id.radioButtonHex -> {
                        updateInputText(Notation.HEX)
                        notation = Notation.HEX
                    }
                    R.id.radioButtonDec -> {
                        updateInputText(Notation.DEC)
                        notation = Notation.DEC
                    }
                    R.id.radioButtonOct -> {
                        updateInputText(Notation.OCT)
                        notation = Notation.OCT
                    }
                    R.id.radioButtonBin -> {
                        updateInputText(Notation.BIN)
                        notation = Notation.BIN
                    }
                }
                initFragment(notation)
            }

            //first checked DEC
            radioButtonDec.isChecked = true
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        if (savedInstanceState != null) {
            expressionText.text = savedInstanceState.getString(KEY_EXPRESSION)
            notation = Notation.valueOf(savedInstanceState.getString(KEY_NOTATION))
            operation = Operation.valueOf(savedInstanceState.getString(KEY_OPERATION))
            curNumber = BigDecimal(savedInstanceState.getString(KEY_NUMBER))
        } else {
            curNumber = BigDecimal.ZERO
            notation = Notation.DEC
            operation = Operation.NONE
        }

        initFragment(Notation.DEC)

        if (resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE) {
            initRadioButtons()
        }

        inputText.setSelection(1)
    }

    private fun initFragment(notation: Notation) {
        var fragment: Fragment? = null
        when (notation) {
            Notation.BIN -> fragment = FragmentBin()
            Notation.OCT -> fragment = FragmentOct()
            Notation.DEC -> fragment = FragmentDec()
            Notation.HEX -> fragment = FragmentHex()
        }
        val transaction = supportFragmentManager.beginTransaction()
        transaction.replace(R.id.fragmentCont, fragment)
                .commit()
    }

    companion object {
        val RESULT_CODE = 200

        val SYMBOL_PLUS = " + "
        val SYMBOL_MINUS = " - "
        val SYMBOL_MULTIPLY = " * "
        val SYMBOL_DEVIDE = " / "
        val SYMBOL_POWER = " ^ "

        val KEY_EXPRESSION = "expression"
        val KEY_NOTATION = "notation"
        val KEY_OPERATION = "operation"
        val KEY_NUMBER = "number"
    }
}
